v-next/hardhat/src/internal/cli/telemetry/sentry/reporter.ts

```diff
@@ -0,0 +1,120 @@
+import {
+  HardhatError,
+  HardhatPluginError,
+} from "@ignored/hardhat-vnext-errors";
+
+import { getHardhatVersion } from "../../../utils/package.js";
+import { isTelemetryAllowed } from "../telemetry-permissions.js";
+
+import { getSubprocessTransport } from "./transport.js";
+
+// TODO: replace with PROD version
+// export const SENTRY_DSN =
+//   "https://38ba58bb85fa409e9bb7f50d2c419bc2@o385026.ingest.sentry.io/522486955555555555";
+export const SENTRY_DSN =
+  "https://d578a176729662a28e7a8da268d36912@o385026.ingest.us.sentry.io/4507685793103872"; // DEV
+
+export async function sendErrorTelemetry(error: Error): Promise<boolean> {
+  const instance = await Reporter.getInstance();
+  return instance.reportError(error);
+}
+
+// ATTENTION: this function is exported for testing, do not directly use it in production
+export function _testResetReporter(): void {
+  Reporter.deleteInstance();
+}
+
+class Reporter {
+  // GENERAL EXPLANATION:
+  // 1) The 'reportError' function collects the error and passes it to our custom Sentry transporter.
+  // 2) The custom transporter receives the JavaScript error serialized by Sentry.
+  // 3) This serialized error is then passed to a detached subprocess, which anonymizes all the information before sending it to Sentry.
+
+  static #instance: Reporter | undefined;
+  readonly #telemetryEnabled: boolean;
+
+  private constructor(telemetryAllowed: boolean) {
+    this.#telemetryEnabled = telemetryAllowed;
+  }
+
+  public static async getInstance(): Promise<Reporter> {
+    if (this.#instance !== undefined) {
+      return this.#instance;
+    }
+
+    const telemetryAllowed = await isTelemetryAllowed();
+    this.#instance = new Reporter(telemetryAllowed);
+
+    const { Integrations, init, setExtra } = await import("@sentry/node");
+
+    const linkedErrorsIntegration = new Integrations.LinkedErrors({
+      key: "parent",
+    });
+
+    const transport = await getSubprocessTransport();
+
+    init({
+      dsn: SENTRY_DSN,
+      transport,
+      integrations: () => [linkedErrorsIntegration],
+    });
+
+    setExtra("nodeVersion", process.version);
+    setExtra("hardhatVersion", await getHardhatVersion());
+
+    return this.#instance;
+  }
+
+  public static deleteInstance(): void {
+    // ATTENTION: only for testing
+    this.#instance = undefined;
+  }
+
+  public async reportError(
+    error: Error,
+    verbose: boolean = false,
+    configPath: string = "",
+  ): Promise<boolean> {
+    if (!(await this.canSendTelemetry(error))) {
+      return false;
+    }
+
+    const { captureException, setExtra } = await import("@sentry/node");
+
+    setExtra("verbose", verbose);
+    setExtra("configPath", configPath);
+
+    captureException(error);
+
+    // TODO: where to run it?
+    // await this.#sentry.flush(1000);
```

Comment:
can you link the docs regarding this?

Reply:
The Sentry docs here are not very clear on the usage of flush (or I did not find the right ones). Is it possible to send events right away without queuing them? Not sure. 

- [This link](https://stackoverflow.com/questions/67241239/sentry-capturemessage-without-flush) mentions that "flush" should NOT be used after every call (but it's for go).

- from [this doc](https://docs.sentry.io/platforms/go/guides/http/configuration/transports/), also for go, it looks like that if the transporter is sync, then the event will be sent in a sync way. But it's a guess. Also, does it apply to NodeJS too?

- [link to JS custom transporter doc](https://docs.sentry.io/platforms/javascript/configuration/transports/)